crates/chain/src/keychain/txout_index.rs

```diff
@@ -3,16 +3,95 @@ use crate::{
     indexed_tx_graph::Indexer,
     miniscript::{Descriptor, DescriptorPublicKey},
     spk_iter::BIP32_MAX_INDEX,
-    SpkIterator, SpkTxOutIndex,
+    DescriptorExt, SpkIterator, SpkTxOutIndex,
+};
+use bitcoin::{
+    hashes::{hash_newtype, sha256},
+    OutPoint, Script, Transaction, TxOut, Txid,
 };
-use bitcoin::{OutPoint, Script, Transaction, TxOut, Txid};
 use core::{
     fmt::Debug,
     ops::{Bound, RangeBounds},
 };
 
 use crate::Append;
 
+hash_newtype! {
+    /// Represents the ID of a descriptor, defined as the sha256 hash of
+    /// the descriptor string, checksum included.
+    pub struct DescriptorId(pub sha256::Hash);
+}
```

Comment:
:100: Nice this is a better solution.

:pick: I'd feel more comfortable with this being defined in `descriptor_ext.rs` (but export it from root). It might get used elsewhere.

:pick: Explaining why this exists would be handy. We want a unique id for a descriptor with a fixed length since descriptors with taproot have no size bound. When we want to persist some application state change associated with the descriptor we don't want to re-write the whole descriptor each time.

:wrench: I think it'd be better if we did not include the checksum in the hash. Is it tricky to get the string representation without the checksum? I wonder if @apoelstra or @sanket1729 has an opinion on this point or the idea of a "descriptor id" in general. Maybe it could be upstreamed.

Reply:
❓ @LLFourn I'm curious why you want to exclude the checksum from the descriptor string hashed for `DescriptorId`. It's easy to remove the checksum by splitting the descriptor string on the `#` but I don't see any reason to do it.

Reply:
@notmandatory I think it's easier to reason about. It's odd to hash a checksum and the checksum is not mandatory for a miniscript descriptor so we should have some standard as to how to come up with a unique identifier for a descriptor (so might as well remove the checksum for this as opposed to keeping it to generate the unique identifier).

Reply:
Ok that makes sense. But the `descriptor_id()` function should already normalizes optional features (including using ' or h and root keys paths) by hashing the `to_string()` of the Descriptor type. But I don't want to bike shed this too much, it's fine for me either way as long as it's done in the `descriptor_id()` function and we use that everywhere.

Reply:
I agree that it's odd to use the hash of the descriptor with the checksum included. I addressed everything in 477c4a0a201296d88b5dcc669bef6b69ecb6e3bb